query-parser/src/parse.rs

```diff
@@ -70,6 +69,130 @@ macro_rules! matches_plain_symbol {
     }
 }
 
+trait FromValue<T> {
+    fn from_value(edn::Value) -> Option<T>;
+}
+
+impl FromValue<PatternNonValuePlace> for PatternNonValuePlace {
+    fn from_value(v: edn::Value) -> Option<PatternNonValuePlace> {
+        match v {
+            edn::Value::Integer(x) => if x >= 0 {
+                Some(PatternNonValuePlace::Entid(x as u64))
+            } else {
+                None
+            },
+            edn::Value::PlainSymbol(x) => if x.0.as_str() == "_" {
+                Some(PatternNonValuePlace::Placeholder)
+            } else if x.is_var_symbol() {
+                Some(PatternNonValuePlace::Variable(Variable(x)))
+            } else {
+                None
+            },
+            edn::Value::NamespacedKeyword(x) =>
+                Some(PatternNonValuePlace::Ident(x)),
+            _ => None,
+        }
+    }
+}
+
+impl FromValue<PatternValuePlace> for PatternValuePlace {
+    fn from_value(v: edn::Value) -> Option<PatternValuePlace> {
+        match v {
+            edn::Value::Integer(x) => Some(PatternValuePlace::EntidOrInteger(x)),
+            edn::Value::PlainSymbol(x) => if x.0.as_str() == "_" {
+                Some(PatternValuePlace::Placeholder)
+            } else if x.is_var_symbol() {
+                Some(PatternValuePlace::Variable(Variable(x)))
+            } else {
+                None
+            },
+            edn::Value::NamespacedKeyword(x) =>
+                Some(PatternValuePlace::IdentOrKeyword(x)),
+            edn::Value::Boolean(x) => Some(PatternValuePlace::Constant(NonIntegerConstant::Boolean(x))),
+            edn::Value::Float(x) => Some(PatternValuePlace::Constant(NonIntegerConstant::Float(x))),
+            edn::Value::BigInteger(x) => Some(PatternValuePlace::Constant(NonIntegerConstant::BigInteger(x))),
+            edn::Value::Text(x) => Some(PatternValuePlace::Constant(NonIntegerConstant::Text(x))),
+            _ => None,
+        }
+    }
+}
+
+impl<I> Query<I>
+    where I: Stream<Item = edn::Value>
+{
+    fn source_var() -> ResultParser<SrcVar, I> {
```

Comment:
I filed https://github.com/mozilla/mentat/issues/218 to extract this `fn_parser` pattern into our macro library.